internal/services/authorization/role_assignment_resource_base.go

```diff
@@ -0,0 +1,383 @@
+			} else if v, ok := metadata.ResourceData.GetOk("role_definition_name"); ok {
+				roleName := v.(string)
+				roleDefinitions, err := roleDefinitionsClient.List(ctx, commonids.NewScopeID(scope), roledefinitions.ListOperationOptions{Filter: pointer.To(fmt.Sprintf("roleName eq '%s'", roleName))})
+				if err != nil {
+					return fmt.Errorf("loading Role Definition List: %+v", err)
+				}
+
+				if roleDefinitions.Model != nil && len(*roleDefinitions.Model) == 1 && (*roleDefinitions.Model)[0].Id != nil {
+					roleDefinitionId = *(*roleDefinitions.Model)[0].Id
+				} else {
+					return fmt.Errorf("loading Role Definition List: failed to find role '%s'", roleName)
+				}
+			} else {
+				return fmt.Errorf("either 'role_definition_id' or 'role_definition_name' needs to be set")
+			}
+
+			metadata.ResourceData.Set("role_definition_id", roleDefinitionId)
+
+			principalId := metadata.ResourceData.Get("principal_id").(string)
+
+			if name == "" {
+				uuid, err := uuid.GenerateUUID()
+				if err != nil {
+					return fmt.Errorf("generating UUID for Role Assignment: %+v", err)
+				}
+
+				name = uuid
+			}
+
+			tenantId := ""
+			delegatedManagedIdentityResourceID := metadata.ResourceData.Get("delegated_managed_identity_resource_id").(string)
+			if len(delegatedManagedIdentityResourceID) > 0 {
+				var err error
+				tenantId, err = getTenantIdBySubscriptionId(ctx, subscriptionClient, subscriptionId)
+				if err != nil {
+					return err
+				}
+			}
+
+			id := parse.NewScopedRoleAssignmentID(scope, name, tenantId)
+			options := roleassignments.DefaultGetOperationOptions()
+			if tenantId != "" {
+				options.TenantId = &tenantId
+			}
+
+			existing, err := roleAssignmentsClient.Get(ctx, id.ScopedId, options)
+			if err != nil {
+				if !response.WasNotFound(existing.HttpResponse) {
+					return fmt.Errorf("checking for presence of existing %s: %s", id, err)
+				}
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return tf.ImportAsExistsError(resourceName, id.ID())
+			}
+
+			properties := roleassignments.RoleAssignmentCreateParameters{
+				Properties: roleassignments.RoleAssignmentProperties{
+					RoleDefinitionId: roleDefinitionId,
+					PrincipalId:      principalId,
+					Description:      pointer.To(metadata.ResourceData.Get("description").(string)),
+				},
+			}
+
+			if len(delegatedManagedIdentityResourceID) > 0 {
+				properties.Properties.DelegatedManagedIdentityResourceId = &delegatedManagedIdentityResourceID
+			}
+
+			condition := metadata.ResourceData.Get("condition").(string)
+			conditionVersion := metadata.ResourceData.Get("condition_version").(string)
+
+			if condition != "" && conditionVersion != "" {
+				properties.Properties.Condition = &condition
+				properties.Properties.ConditionVersion = &conditionVersion
+			} else if condition != "" || conditionVersion != "" {
+				return fmt.Errorf("`condition` and `conditionVersion` should be both set or unset")
+			}
```

Comment:
Won't the `RequiredWith` in the schema already validate this/prevent this scenario from happening?

Reply:
yes, duplicated. Code updated.